IdentityCore/src/cache/metadata/MSIDMetadataCache.m

```diff
@@ -0,0 +1,125 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSIDMetadataCache.h"
+#import "MSIDCache.h"
+#import "MSIDCacheItemJsonSerializer.h"
+#import "MSIDJsonSerializer.h"
+#import "MSIDJsonSerializing.h"
+#import "MSIDCacheKey.h"
+#import "MSIDAccountMetadataCacheItem.h"
+#import "MSIDAccountMetadataCacheKey.h"
+
+@implementation MSIDMetadataCache
+{
+    NSMutableDictionary *_memoryCache;
+    id<MSIDMetadataCacheDataSource> _dataSource;
+    dispatch_queue_t _synchronizationQueue;
+    MSIDCacheItemJsonSerializer *_jsonSerializer;
+}
+
+- (instancetype)initWithPersistantDataSource:(id<MSIDMetadataCacheDataSource>)dataSource
+{
+    if (!dataSource) return nil;
+    
+    self = [super init];
+    
+    if (self)
+    {
+        _memoryCache = [NSMutableDictionary new];
+        _dataSource = dataSource;
+        NSString *queueName = [NSString stringWithFormat:@"com.microsoft.msidmetadatacache-%@", [NSUUID UUID].UUIDString];
+        _synchronizationQueue = dispatch_queue_create([queueName cStringUsingEncoding:NSASCIIStringEncoding], DISPATCH_QUEUE_CONCURRENT);
+        _jsonSerializer = [MSIDCacheItemJsonSerializer new];
+    }
+    
+    return self;
+}
+
+- (BOOL)saveAccountMetadata:(MSIDAccountMetadataCacheItem *)item
+                        key:(MSIDAccountMetadataCacheKey *)key
+                    context:(id<MSIDRequestContext>)context
+                      error:(NSError **)error
+{
+    if (!item || key)
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain,
+                                     MSIDErrorInvalidInternalParameter,
+                                     @"cacheItem and key could not be nil.",
+                                     nil, nil, nil, nil, nil);
+        }
+        return NO;
+    }
+    
+    __block NSError *localError;
+    __block BOOL saveSuccess = NO;
+    
+    dispatch_barrier_sync(_synchronizationQueue, ^{
+        MSIDAccountMetadataCacheItem *currentItem = [_dataSource accountMetadataWithKey:key serializer:_jsonSerializer context:context error:&localError];
```

Comment:
I'm curious, what would be better for performance:
1. Reading item, comparing if it's the same, and updating only if it changed (1 read always + 1 potential write)
2. Always updating (1 write always)

Write is probably more costly than read, but considering we don't know how exactly keychain implements both of them, would be a curious discussion to have.

Reply:
Let's do this: 
1. check in memory, and if same, then skip saving.
2. for other cases, let's just save.